sidebar-row: Show mini thumbnails

marhkb opened the pull request:
Fixes #297

marhkb committed:
sidebar-row: Separate sender label from content

This will help to show mini thumbnails for photo messages.

melix99 commented on src/session/sidebar/row.rs:
```diff
@@ -359,9 +370,78 @@ impl Row {
+                                                target_size,
+                                                target_size,
+                                                gdk_pixbuf::InterpType::Bilinear,
+                                            )
```
I'd really avoid doing that expensive operation honestly. Also I don't really want to use `GdkPixbuf` as that API is meant to go away in the future. What we need here is something similar to css's [object-fit: cover](https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit) which isn't available natively in GTK widgets (I know that there was some discussion to add that, but no action was taken). So... we have to implement our own widget for now.

Fortunately, the `MediaPicture` widget is pretty close to what we need. You just need to remove the `aspect-ratio` property, since we need a `1:1` widget, then you should use the paintable intrinsic aspect ratio inside the snapshot virtual function. Also, we don't really need the `measure` and `request_mode` thing, since we just need to set minimum sizes via `width-request` and `height-request`.

melix99 commented on src/session/sidebar/row.rs:
```diff
@@ -359,9 +370,78 @@ impl Row {
+                                                target_size,
+                                                target_size,
+                                                gdk_pixbuf::InterpType::Bilinear,
+                                            )
```
I'd really avoid doing that expensive operation honestly. Also I don't really want to use `GdkPixbuf` as that API is meant to go away in the future. What we need here is to something similar to css's [object-fit: cover](https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit) which isn't available natively in GTK widgets (I know that there was some discussion to add that, but no action was taken). So... we have to implement our own widget for now.

Fortunately, the `MediaPicture` widget is pretty close to what we need. You just need to remove the `aspect-ratio` property, since we need a `1:1` widget, then you should use the paintable intrinsic aspect ratio inside the snapshot virtual function. Also, we don't really need the measure and request_mode thing, since we just need to set minimum sizes via `width-request` and `height-request`.

melix99 replied:
GitHub is drunk and it decided to post a double comment.

marhkb committed:
Introduce widget `SidebarRowMiniThumbnail`

This widget will be used to show mini thumbnails for specific messages in the sidebar.

marhkb committed:
sidebar-row: Show mini thumbnails

Fixes #297

Co-authored-by: Marco Melorio <marco.melorio@protonmail.com>

melix99 commented on src/session/sidebar/mini_thumbnail.rs:
```diff
@@ -0,0 +1,99 @@
+                    snapshot.upcast_ref(),
+                    widget.width() as f64,
+                    widget.height() as f64,
+                );
```
This way the image will appear stretched when it's not perfectly squared. For instance (Telegrand on the left):

![image](https://user-images.githubusercontent.com/23530586/172501802-553bbaa4-5958-44d4-8ada-3780e994b6be.png)

This is basically `object-fit: fill`, but we need `cover`. The snapshot implementation in `MediaPicture` basically does that, so you can just use that.

marhkb replied:
Ah, okay. I always thought the mini thumbnails were kind of stretched. Next try, I guess.

melix99 commented on src/session/sidebar/mini_thumbnail.rs:
```diff
@@ -0,0 +1,158 @@
+                let widget_width = widget.width() as f64;
+                let widget_height = widget.height() as f64;
+                let widget_ratio = widget_width / widget_height;
+                let paintable_ratio = widget.aspect_ratio();
```
```suggestion
                let paintable_ratio = paintable.instinsic_aspect_ratio();
```

There's no need to have a separate property here. This is different in the `MediaPicture` because we wanted to size the widget even when the paintable was not ready (e.g. because it was downloading), but in this case we always have a paintable, so it's not a problem.

melix99 commented on src/session/sidebar/mini_thumbnail.rs:
```diff
@@ -0,0 +1,158 @@
+            obj.set_overflow(gtk::Overflow::Hidden);
+            obj.set_height_request(16);
+            obj.set_width_request(16);
+            obj.add_css_class("mini-thumbnail");
```
This should be defined in the `class_init` function. Also, by convention the css name shouldn't have any dash.

melix99 replied:
Oh, I didn't notice you were using a css class. Then, I suggest to set a css name in `class_init` instead. Similar to what I did [here](https://github.com/melix99/telegrand/blob/95a07e003d9ca3142eddab8249827a233c77b825/src/session/content/message_row/indicators.rs#L56).

melix99 commented on data/resources/ui/sidebar-row.ui:
```diff
@@ -64,15 +64,37 @@
+                <child>
+                  <object class="SidebarMiniThumbnail" id="message_thumbnail">
+                    <property name="margin-start">2</property>
+                    <property name="margin-end">2</property>
```
```suggestion
```

Actually I don't mind having them in-line with the text.

melix99 commented on src/session/sidebar/row.rs:
```diff
@@ -131,6 +136,19 @@ mod imp {
+
+            self.message_prefix_label.connect_visible_notify(clone!(@weak obj => move |label|
+                obj.imp().message_thumbnail.set_margin_start(if label.is_visible() { 2 } else { 0 });
+            ));
```
```suggestion
```

Same as above ^.
